Rename WebCore::findHTMLTag to findTag and extend it to find SVG and MathML tags

cdumez opened the pull request:
#### 67070d4e441692ab2cda8a9fe5257006742ff0da
<pre>
Rename WebCore::findHTMLTag to findTag and extend it to find SVG and MathML tags
<a href="https://bugs.webkit.org/show_bug.cgi?id=244020">https://bugs.webkit.org/show_bug.cgi?id=244020</a>
rdar://problem/98767683

Reviewed by NOBODY (OOPS!).

We currently generate the WebCore::findHTMLTag function and use this in the
AtomHTMLToken constructor. We can extend this to look for any known tag name
(HTML, SVG, or MathML) by reworking make-names.pl a bit.

This isn&apos;t helpful for Speedometer performance, since parsing non-HTML tags is
rare, but local testing shows it to be a neutral change. Handling all known tag
names will set the stage for a later patch that does improve Speedometer
performance.

We generate a new pair of files, KnownTag.h and KnownTag.cpp, move the
old findHTMLTag function there, and rename it to findTag.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreMacros.cmake:

Build system changes to invoke make_names.pl to generate
KnownTag.{h,cpp}.

* Source/WebCore/dom/make_names.pl:
(readKnownTags):

Read a list of tag name files. Any duplicate tag name found in a
subsequent file is ignored (e.g. the SVG script tag will be ignored if
the HTML tag name file is parsed first). Since the %parameters hash
doesn&apos;t make much sense when parsing multiple files, it&apos;s cleared out
by the end of the function, but the namespace for each tag (which comes
from the parameters object) is recorded on the tag object.

(printNamesHeaderFile):
(printNamesCppFile):

Remove the findHTMLTag generation from generated HTMLNames.{h,cpp} etc.

(printKnownTagsHeaderFile):
(printKnownTagsCppFile):

Generated findTag in KnownTag.{h,cpp}. Use the namespace stored on the
tag object to decide the QualifiedName on which to look up the local
name.

(findMaxTagLength): Just refer to the global %allTags variable, like other
functions do.

(tagsWithLength): Simplify.

(generateFindTagForLength): We need to replace &quot;_&quot;s with &quot;-&quot;s when generating
the big switch statement, since the tag names it&apos;s operating on have already
been converted into identifiers.

* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::AtomHTMLToken):

Call the new findTag function.
</pre>

darinadler commented on Source/WebCore/Modules/mediasource/MediaSourceRegistry.h:
```diff
@@ -47,13 +47,15 @@ class MediaSourceRegistry final : public URLRegistry {
 private:
     MediaSourceRegistry();
-    MemoryCompactRobinHoodHashMap<String, RefPtr<MediaSource>> m_mediaSources;
+    MemoryCompactRobinHoodHashMap<String, std::pair<RefPtr<MediaSource>, ScriptExecutionContextIdentifier>> m_mediaSources;
```
Can the pairs use `Ref` instead of 15th RefPtr`?

cdumez replied:
I'll try. I'm not quite sure how it works when you call HashMap::get() and want to make sure the value is valid (i.e. the key was in the map). Currently we null check the media source but if it becomes a Ref then that no longer works.

cdumez replied:
Oh, it seems I can use Ref<> and then check Ref<>::isHashTableEmptyValue() after calling HashMap::get() / HashMap::take().

darinadler replied:
I sort of regret suggesting it if that's how we have to use it. But either way is of course OK.

darinadler commented on Source/WebCore/html/URLRegistry.h:
```diff
@@ -46,9 +47,14 @@ class URLRegistrable {
-    virtual ~URLRegistry() = default;
-    virtual void registerURL(ScriptExecutionContext&, const URL&, URLRegistrable&) = 0;
-    virtual void unregisterURL(const URL&) = 0;
+    static Vector<URLRegistry*> allRegistries();
```
I probably would have had this return something that iterates across references rather than a vector of pointers, just to make the caller code look nicer. But I suppose that would still require a `Vector` underneath it, or maybe a `FixedVector` since we never need to modify it. Another alternative is to just expose a `forEach` function instead.

cdumez replied:
Ok, I'll look into a forEach() function. I hesitated going this way. I eventually went with the Vector because it allowed the call site to easily stop iterating when finding the right registry. That said, there are only 2 registries at the moment so not an important optimization.

darinadler commented on Source/WebCore/html/URLRegistry.cpp:
```diff
@@ -0,0 +1,63 @@
+
+URLRegistry::~URLRegistry()
+{
+    RELEASE_ASSERT_NOT_REACHED(); // All our registries are singleton objects.
```
Wish we could delete the destructor instead of implementing it. Would be nicer to get a compile or link time error if we destroy by accident.

cdumez replied:
It would be nice but it's not possible here right? We want to use polymorphism so we need the base class to have a destructor that is virtual.

cdumez replied:
Looks like clang at least is letting me delete the destructor. It is usually not good practice to not have a virtual destructor when having virtual functions (and I thought the compiler would warn about that) but it should be OK here given that this is never deleted. Should I do it?

darinadler replied:
You probably should not. I'd think the destructor would be needed for the vtable used when constructing an object of this class, for example, even if it's never called.

darinadler commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -55,21 +57,56 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(Blob);
-void BlobURLRegistry::unregisterURL(const URL& url)
+bool BlobURLRegistry::unregisterURL(const URL& url)
 {
+    bool isURLRegistered = false;
```
`wasURLRegistered`

cdumez replied:
Maybe it's my bad English but `wasURLRegistered` sounds a little more confusing to me than `isURLRegistered` in this context. First we check if the URL *is* registered, and if it is registered then I unregister it. 

If this were a registerURL() function and I'd want a boolean to keep track of whether or not the function actually registered the URL, then I would use `wasURLRegistered` because it would convey: "Did the earlier code register the URL".

As I said, it may just be my English.

darinadler commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -55,21 +57,56 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(Blob);
     static URLRegistry& registry();
+
+    Lock m_urlsPerContextLock;
+    HashMap<ScriptExecutionContextIdentifier, HashSet<URL>> m_urlsPerContext WTF_GUARDED_BY_LOCK(m_urlsPerContextLock);
```
We might want to make a type for maps of sets this includes the removal algorithm that handles the set becoming empty. Might even optimize that case of removing the last item in the set so we don’t first remove from the hash table and then destroy it.

cdumez replied:
I think this would be nice (though likely separately from this patch). I have definitely written such logic quite a few times so having something reusable would be nice. However, it may need to be templated because the nested type is not always a HashSet<>. I remember writing code where the inner type was a Vector for example.

cdumez committed:
Support revoking Blob URLs across same-origin contexts
https://bugs.webkit.org/show_bug.cgi?id=243905

Reviewed by Darin Adler.

Support revoking Blob URLs across same-origin contexts, as per the specification:
- https://w3c.github.io/FileAPI/#dfn-revokeObjectURL

This also aligns our behavior with Blink and Firefox.

Previously, only the context that created the blob URL could revoke it.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/cross-global-revoke.sub-expected.txt:
Rebaseline WPT test that is now passing.

* LayoutTests/imported/w3c/web-platform-tests/FileAPI/url/cross-global-revoke.sub.html:
Fix test so that the 3rd subtest is able to run with our infrastructure:
https://github.com/web-platform-tests/wpt/pull/35455

* Source/WebCore/Modules/mediasource/MediaSourceRegistry.cpp:
(WebCore::MediaSourceRegistry::registerURL):
(WebCore::MediaSourceRegistry::unregisterURL):
(WebCore::MediaSourceRegistry::unregisterURLsForContext):
(WebCore::MediaSourceRegistry::lookup const):
* Source/WebCore/Modules/mediasource/MediaSourceRegistry.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::BlobURLRegistry::registerURL):
(WebCore::BlobURLRegistry::unregisterURL):
(WebCore::BlobURLRegistry::unregisterURLsForContext):
* Source/WebCore/html/PublicURLManager.cpp:
(WebCore::PublicURLManager::PublicURLManager):
(WebCore::PublicURLManager::registerURL):
(WebCore::PublicURLManager::revoke):
(WebCore::PublicURLManager::stop):
* Source/WebCore/html/PublicURLManager.h:
* Source/WebCore/html/URLRegistry.cpp: Added.
(WebCore::URLRegistry::URLRegistry):
(WebCore::URLRegistry::~URLRegistry):
* Source/WebCore/html/URLRegistry.h:

Canonical link: https://commits.webkit.org/253416@main
